### PR TITLE
[RN][Docs] Add guide on custom types in C++

### PR DESCRIPTION
**Pull request by cipolleschi**
This PR depends on #4254. 

This adds the advanced use cases to bridge custom types from JS to C++ and viceversa

### CODE REVIEW

File: docs/the-new-architecture/custom-cxx-types.md [hunk 1]
```diff
@@ -0,0 +1,417 @@
...
+```
+
+The key components for your custom bridging header are:
+
+- Explicit specialization of the `Bridging` struct for your custom type. In this case, the template specify the `int64_t` type.
+- A `fromJs` function to convert from the JS representation to the C++ representation
+- A `toJs` function to convert from the C++ representation to the JS representation
+
+:::note
+On iOS, remember to add the `Int64.h` file to the Xoced project.
```
Comment (cortinico) [inline docs/the-new-architecture/custom-cxx-types.md:64]: typo

File: docs/the-new-architecture/custom-cxx-types.md [hunk 2]
```diff
@@ -0,0 +1,417 @@
...
+:::
+
+### 2. Modify the JS Spec
+
+Now, we can modify the JS spec to add a method that uses the new type. As usual, we can use either Flow or TypeScript for our specs.
+
+1. Open the `specs/NativeSampleTurbomodule`
+2. Modify the spec as it follows:
```
Comment (cortinico) [inline docs/the-new-architecture/custom-cxx-types.md:72]:
```suggestion
2. Modify the spec as follows:
```

File: docs/the-new-architecture/custom-cxx-types.md [hunk 3]
```diff
@@ -0,0 +1,417 @@
...
+
+1. Open the `specs/NativeSampleModule.h` file and apply the following changes:
+
+```diff title="NativeSampleModule.h"
+#pragma once
+
+#include <AppSpecsJSI.h>
++ #include "Int64.h"
+#include <memory>
+#include <string>
```
Comment (cortinico) [inline docs/the-new-architecture/custom-cxx-types.md:125]: can we sort them and have all the diamon import before/after the quote includes?

File: docs/the-new-architecture/custom-cxx-types.md [hunk 4]
```diff
@@ -0,0 +1,417 @@
...
++export type Address = {
++  street: string,
++  num: number,
++  isInUS: boolean,
++};
+
+
+export interface Spec extends TurboModule {
+  +reverseString: (input: string) => string;
++  +validateAddress: (input: Address) => boolean;
```
Comment (cortinico) [inline docs/the-new-architecture/custom-cxx-types.md:271]:
```suggestion
+ +validateAddress: (input: Address) => boolean;
```

File: docs/the-new-architecture/custom-cxx-types.md [hunk 5]
```diff
@@ -0,0 +1,417 @@
...
+
++export type Address = {
++  street: string,
++  num: number,
++  isInUS: boolean,
++};
+
+export interface Spec extends TurboModule {
+  readonly reverseString: (input: string) => string;
++  readonly validateAddress: (input: Address) => boolean;
```
Comment (cortinico) [inline docs/the-new-architecture/custom-cxx-types.md:245]:
```suggestion
+ readonly validateAddress: (input: Address) => boolean;
```

File: docs/the-new-architecture/custom-cxx-types.md [hunk 6]
```diff
@@ -0,0 +1,417 @@
...
+</TabItem>
+</Tabs>
+
+This code defines the new `Address` type and defines a new `validateAddress` function for the Turbo Native Module. Notice that the `validateFunction` requires an `Address` object as parameter.
+
+It is also possible to have functions that return custom types.
+
+### 2. Define the bridging code
+
+From the Address type defined in the specs, Codegen will generate two helper types: `NativeSampleModuleAddress` and `NativeSampleModuleAddressBridging`.
```
Comment (cortinico) [inline docs/the-new-architecture/custom-cxx-types.md:288]:
```suggestion
From the `Address` type defined in the specs, Codegen will generate two helper types: `NativeSampleModuleAddress` and `NativeSampleModuleAddressBridging`.
```

File: docs/the-new-architecture/custom-cxx-types.md [hunk 7]
```diff
@@ -0,0 +1,417 @@
...
+
+The first type is the definition of the `Address`. The second type contains all the plumbing to bridge the custom type from JS to C++ and viceversa. The only extra step we need to add is to define the `Bridging` structure that extends the `NativeSampleModuleAddressBridging` type.
```
Comment (cortinico) [inline docs/the-new-architecture/custom-cxx-types.md:290]: all the plumbing -> all the infrastructure

File: docs/the-new-architecture/custom-cxx-types.md [hunk 8]
```diff
@@ -0,0 +1,417 @@
...
+This code defines an `Address` typealias for the generic type `NativeSampleModuleAddress`. **The order of the generics matters**: the first template argument refers to the first data type of the struct, the second refers to the second, and so forth.
+
+Then, the code adds the `Bridging` specialization for the new `Address` type, by extending `NativeSampleModuleAddressBridging` that is generated by Codegen.
+
+:::note
+There is a convention that is followed to generate this types:
+
+- The first part of the name is always the type of the module. `NativeSampleModule`, in this example.
+- The second part of the name is always the name of the JS type defined in the specs. `Address`, in this example.
+  :::
```
Comment (cortinico) [inline docs/the-new-architecture/custom-cxx-types.md:319]:
```suggestion
:::
```

File: docs/the-new-architecture/custom-cxx-types.md [hunk 9]
```diff
@@ -0,0 +1,417 @@
...
+```c++ title="NativeSampleModule.cpp (validateAddress implementation)"
+bool NativeSampleModule::validateAddress(jsi::Runtime &rt, jsi::Object input) {
+  std::string street = input.getProperty(rt, "street").asString(rt).utf8(rt);
+  int32_t number = input.getProperty(rt, "num").asNumber();
+
+  return !street.empty() && number > 0;
+}
+```
+
+In the implementation, the object that represents the `Address` is a `jsi::Object`. To extract the values from this object, we need to use the accessors provided by `JSI` (If you want to learn more about `JSI`, have a look at this [great talk](https://youtu.be/oLmGInjKU2U?feature=shared) from App.JS 2024):
```
Comment (cortinico) [inline docs/the-new-architecture/custom-cxx-types.md:347]: Not sure I would link the talk here. Maybe in a section "Learn more" at the bottom?

Comment (cipolleschi) [reply docs/the-new-architecture/custom-cxx-types.md:347]: This is the only point in the whole document where we talk about JSI. We can definitely add a `Learn More` at the bottom of the page, but it feels decontextualized to me...

What if we add it right below this explanation? PErhaps in a `note` box

Comment (cortinico) [reply docs/the-new-architecture/custom-cxx-types.md:347]: Yeah the note box also works

File: docs/the-new-architecture/custom-cxx-types.md (hunk 8 above)
Comment (cortinico) [reply docs/the-new-architecture/custom-cxx-types.md:320]: This was not committed actually

**Review by cortinico — APPROVED** [review]

**Pull request merged**